Move to Travis CI for building releases/nightlies

adriaanm opened the pull request.

adriaanm committed:
Switch to Travis CI for building releases

Both nightly and full releases migrate to Travis.
PR validation remains on our Jenkins cluster for now.

Main difference is that we don't use our artifactory as
a cache, nor do we need to wipe stuff, since Travis
gives us a fresh machine. We do cache the ivy/sbt cache.

TODO: decide whether to publish a "mergely" (a nightly
for each merge, or skip this script on merge and use
a scheduled job for publishing the nightly.)

Note: we don't use `travis encrypt-file` because it
nukes the iv/key variables on each invocation.. too much magic

Instead, I did:

```
cat /dev/urandom | head -c 10000 | openssl sha1 > ./secret
openssl aes-256-cbc -pass "file:./secret" -in gpg_subkey -out admin/files/gpg_subkey.enc
travis encrypt "GPG_SUBKEY_SECRET=$(cat ./secret)"
```

adriaanm committed:
Pull structure of bootstrap script into .travis.yml,...

so that:
  - we can use build stages
  - we get a nicer experience with build logs
  - we can easily see / tweak the overall flow with env vars

Currently, we are skipping the run tests (take too long),
and stability (broken until Stefan's PR is merged?).

To publish a build, trigger a custom build with a before_script
that set these env variables:
  - `SCALA_VER_BASE`
  - `SCALA_VER_SUFFIX`
  - `publishToSonatype` (the yaml looks at this to skip the step,
    so it can't be set by our scripts I think)

(TODO: let's just set `SCALA_VER`)

adriaanm committed:
Exclude PR validation as a trigger

adriaanm commented:
Hmm, wait -- how is the event type `push`?? Doesn't this mean anyone could submit a PR to reveal secrets??

lrytz commented on admin/init.sh:
```diff
@@ -0,0 +1,30 @@
+
+# directories needed by sensitive part
+# mkdir -p ~/.m2 -- not needed anymore (used for ide integration?)
+mkdir -p ~/.ssh
```
I think ssh complains if the directory has default permissions, but we're probably lucky here because it already exists. https://superuser.com/questions/215504/permissions-on-private-key-in-ssh-folder. It seems my `chmod` in scala-dist (https://github.com/scala/scala-dist/blob/2.12.x/admin/build.sh#L54) is a bit too rough, 600 is not right for all files in .ssh.

Why do we even create the folder here?

adriaanm replied:
Dropped it

lrytz commented on scripts/common:
```diff
@@ -159,19 +159,36 @@ EOF
     done
   fi
+
+  if [ "${TRAVIS}" != "true" ]; then
```
As far as I see, `generateRepositoriesConfig` is only called if not-travis, so is this redundant?

adriaanm replied:
I squirreled [one more invocation](https://github.com/scala/scala/pull/6390/files#diff-f1be5674025c26b5d092359fc6c6d9c4R116) away in `mkFreshIvy`, the repo config is still needed. It would be nice to get rid of it by adding the needed resolvers, but I'm not sure where :-/

adriaanm replied:
Maybe something like:
```
addSbtPlugin("com.jsuereth" % "sbt-pgp" % "1.1.0")

resolvers ++= Seq(Resolver.typesafeIvyRepo("releases"), Resolver.sbtPluginRepo("releases"))
```

adriaanm replied:
We also need that resolver to get STARR. How do I do that?

adriaanm replied:
I'm going to punt on this. Would be really nice if we could avoid creating a repositories config, but I don't know how

adriaanm replied:
Thanks to some help from @szeiger and @dwijnand, I resolved this too :-). Now I'm rewarded with removing this redundant check.

lrytz commented on scripts/jobs/integrate/bootstrap:
```diff
@@ -229,15 +235,15 @@ buildScalaCheck(){
 # build modules, using ${buildTasks[@]} (except for ScalaCheck, which is hard-coded to publish to artifactory)
 buildModules() {
-  publishTasks=('set credentials += Credentials(Path.userHome / ".credentials-private-repo")' "set every publishTo := Some(\"publish-repo\" at \"$integrationRepoUrl\")")
+  publishTasks=('set credentials += Credentials("Artifactory Realm", "scala-ci.typesafe.com", "scala-ci", System.getenv("PRIVATE_REPO_PASS"))' "set every publishTo := Some(\"publish-repo\" at \"$integrationRepoUrl\")")
```
will this still work on jenkins / is `PRIVATE_REPO_PASS` defind there?

lrytz commented on .travis.yml:
```diff
@@ -1,26 +1,73 @@
 
-install: bundle install
+# buildQuick needs following env (is that propagated to stages?)
+# - PRIVATE_REPO_PASS, integrationRepoUrl,
```
is `PRIVATE_REPO_PASS` on jenkins?

lrytz commented on scripts/common:
```diff
@@ -4,23 +4,38 @@ trap "exit 1" TERM
+WORKSPACE="${WORKSPACE-`pwd`}"
+
+# the default (home dir) is fine on Travis, since each jobs gets its own worker (ivy cache is cached by travis)
+IVY_HOME="${IVY_HOME-$HOME/.ivy2}"
```
how / where is this used?

adriaanm replied:
good catch, forgot to remove

lrytz commented on scripts/common:
```diff
@@ -4,23 +4,38 @@ trap "exit 1" TERM
+# SBT_HOME="$WORKSPACE/.sbt"
+# mkdir -p $SBT_HOME
+# IVY_CACHE="$WORKSPACE/.ivy2"
+# mkdir -p $IVY_CACHE
```
I don't get the full picture why we don't need those anymore; I see the `bootstrap` script (for jenkins) uses `sbtArgs` to set `-ivy/-Dsbt.global.base/-sbt-dir`, but it did that before. So the variables were unused before? Can we just remove them?

adriaanm replied:
Yeah, i think they were completely unused for a while. Maybe ide integration used them?

adriaanm replied:
I'm going to leave a trace in the comments for now, because it's so hard to convince yourself you didn't overlook a use site.

adriaanm replied:
I just realized I could use https://github.com/search?q=org%3Ascala+SBT_HOME&type=Code

nuking the comments

lrytz commented on .travis.yml:
```diff
@@ -1,26 +1,73 @@
+          - source scripts/common
+          - source scripts/bootstrap_fun
+          - mkFreshIvy
+          - find build -type f -exec touch {} + # give antStyle a chance
```
That didn't work out so far, right?

adriaanm replied:
Nope

lrytz commented on .travis.yml:
```diff
@@ -14,12 +14,22 @@ cache:
   - (cd admin && ./init.sh)
 
+stages:
+  - name: build
```
maybe add a comment that the `build` stage is not excluded on `pull_request` because it has a (parallel) job for the spec

adriaanm replied:
ok

adriaanm committed:
Review feedback from lrytz

adriaanm commented:
I killed the jenkins jobs, since they're not testing anything relevant.

adriaanm commented:
/nothingtoseehere

adriaanm committed:
Cleanups, skip tests for now

Main cleanup = no more repositories config!
This is nice because we were duplicating the one that
comes with sbt's launcher, just to add a few lines.

We could do the same for the jenkins scripts in scripts/,
but leaving those alone as much as possible for now.

adriaanm committed:
Hopefully last round of cleanups

adriaanm commented:
This is ready for final round of reviews, and merging 🤞

lrytz reviewed (approved):
Excited, dankeschön!

adriaanm commented:
![](https://media1.tenor.com/images/06a3a9c25ed19f40a088bbcda4146a3d/tenor.gif?itemid=8312851)
